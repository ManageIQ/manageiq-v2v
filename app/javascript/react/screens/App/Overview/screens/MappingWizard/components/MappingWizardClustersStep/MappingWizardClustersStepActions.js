import URI from 'urijs';
import API from '../../../../../../../../common/API';
import {
  FETCH_V2V_SOURCE_CLUSTERS,
  FETCH_V2V_TARGET_CLUSTERS,
  REMOVE_TARGET_CLUSTER,
  REMOVE_SOURCE_CLUSTERS
} from './MappingWizardClustersStepConstants';

import {
  requestSourceClustersData,
  requestTargetClustersData
} from './mappingWizardClustersStep.fixtures';

const _getSourceClustersActionCreator = url => dispatch =>
  dispatch({
    type: FETCH_V2V_SOURCE_CLUSTERS,
    payload: API.get(url)
  }).catch(error => {
    // redux-promise-middleware will automatically send:
    // FETCH_V2V_SOURCE_CLUSTERS_PENDING, FETCH_V2V_SOURCE_CLUSTERS_FULFILLED, FETCH_V2V_SOURCE_CLUSTERS_REJECTED

    // to enable UI development without the backend ready, i'm catching the error
    // and passing some mock data thru the FULFILLED action after the REJECTED action is finished.
    dispatch({
      type: `${FETCH_V2V_SOURCE_CLUSTERS}_FULFILLED`,
      payload: requestSourceClustersData.response
    });
  });

export const fetchSourceClustersAction = url => {
  const uri = new URI(url);
  return _getSourceClustersActionCreator(uri.toString());
};

const _getTargetClustersActionCreator = url => dispatch =>
  dispatch({
    type: FETCH_V2V_TARGET_CLUSTERS,
    payload: API.get(url)
  }).catch(error => {
    // redux-promise-middleware will automatically send:
    // FETCH_V2V_TARGET_CLUSTERS_PENDING, FETCH_V2V_TARGET_CLUSTERS_FULFILLED, FETCH_V2V_TARGET_CLUSTERS_REJECTED

    // to enable UI development without the backend ready, i'm catching the error
    // and passing some mock data thru the FULFILLED action after the REJECTED action is finished.
    dispatch({
      type: `${FETCH_V2V_TARGET_CLUSTERS}_FULFILLED`,
      payload: requestTargetClustersData.response
    });
  });

export const fetchTargetClustersAction = url => {
  const uri = new URI(url);
  return _getTargetClustersActionCreator(uri.toString());
};

export const removeTargetCluster = targetClusterToRemove => dispatch =>
  dispatch({
    type: REMOVE_TARGET_CLUSTER,
    targetClusterToRemove
  });

export const removeSourceClusters = sourceClustersToRemove => dispatch =>
  dispatch({
    type: REMOVE_SOURCE_CLUSTERS,
    sourceClustersToRemove
  });
