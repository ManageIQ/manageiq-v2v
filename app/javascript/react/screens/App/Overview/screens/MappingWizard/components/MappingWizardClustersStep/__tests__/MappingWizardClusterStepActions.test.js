import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import * as actions from '../MappingWizardClustersStepActions';
import {
  initialState,
  requestSourceClustersData,
  requestTargetClustersData
} from '../mappingWizardClustersStep.fixtures';
import {
  mockRequest,
  mockReset
} from '../../../../../../../../../common/mockRequests';

const middlewares = [thunk, promiseMiddleware()];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

afterEach(() => {
  store.clearActions();
  mockReset();
});

describe('mappingWizard actions', () => {
  it('should fetch source clusters and return PENDING and FULFILLED action', () => {
    const { fetchSourceClustersUrl } = requestSourceClustersData;
    mockRequest({
      fetchSourceClustersUrl,
      status: 200
    });
    return store
      .dispatch(actions.fetchSourceClustersAction(fetchSourceClustersUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
  it('should fetch source clusters and return PENDING and REJECTED action', () => {
    const { fetchSourceClustersUrl } = requestSourceClustersData;
    mockRequest({
      fetchSourceClustersUrl,
      status: 404
    });
    return store
      .dispatch(actions.fetchSourceClustersAction(fetchSourceClustersUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });

  it('should fetch target clusters and return PENDING and FULFILLED action', () => {
    const { fetchTargetClustersUrl } = requestTargetClustersData;
    mockRequest({
      fetchTargetClustersUrl,
      status: 200
    });
    return store
      .dispatch(actions.fetchTargetClustersAction(fetchTargetClustersUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
  it('should fetch target clusters and return PENDING and REJECTED action', () => {
    const { fetchTargetClustersUrl } = requestTargetClustersData;
    mockRequest({
      fetchTargetClustersUrl,
      status: 404
    });
    return store
      .dispatch(actions.fetchTargetClustersAction(fetchTargetClustersUrl))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
});
