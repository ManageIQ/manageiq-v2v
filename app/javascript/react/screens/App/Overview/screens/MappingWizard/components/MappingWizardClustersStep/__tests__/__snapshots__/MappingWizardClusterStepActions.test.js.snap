// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mappingWizard actions should fetch source clusters and return PENDING and FULFILLED action 1`] = `
Array [
  Object {
    "type": "FETCH_V2V_SOURCE_CLUSTERS_PENDING",
  },
  Object {
    "payload": Object {
      "config": Object {
        "adapter": null,
        "data": undefined,
        "headers": Object {
          "Accept": "application/json, text/plain, */*",
          "X-Auth-Token": "",
          "X-Requested-With": "XMLHttpRequest",
        },
        "maxContentLength": -1,
        "method": "get",
        "params": Object {},
        "timeout": 0,
        "transformRequest": Object {
          "0": [Function],
        },
        "transformResponse": Object {
          "0": [Function],
        },
        "url": "/api/dummyProviders",
        "validateStatus": [Function],
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
      },
      "data": null,
      "headers": undefined,
      "status": 200,
    },
    "type": "FETCH_V2V_SOURCE_CLUSTERS_FULFILLED",
  },
]
`;

exports[`mappingWizard actions should fetch source clusters and return PENDING and REJECTED action 1`] = `
Array [
  Object {
    "type": "FETCH_V2V_SOURCE_CLUSTERS_PENDING",
  },
  Object {
    "error": true,
    "payload": [Error: Request failed with status code 404],
    "type": "FETCH_V2V_SOURCE_CLUSTERS_REJECTED",
  },
]
`;

exports[`mappingWizard actions should fetch target clusters and return PENDING and FULFILLED action 1`] = `
Array [
  Object {
    "type": "FETCH_V2V_TARGET_CLUSTERS_PENDING",
  },
  Object {
    "payload": Object {
      "config": Object {
        "adapter": null,
        "data": undefined,
        "headers": Object {
          "Accept": "application/json, text/plain, */*",
          "X-Auth-Token": "",
          "X-Requested-With": "XMLHttpRequest",
        },
        "maxContentLength": -1,
        "method": "get",
        "params": Object {},
        "timeout": 0,
        "transformRequest": Object {
          "0": [Function],
        },
        "transformResponse": Object {
          "0": [Function],
        },
        "url": "/api/dummyProviders",
        "validateStatus": [Function],
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
      },
      "data": null,
      "headers": undefined,
      "status": 200,
    },
    "type": "FETCH_V2V_TARGET_CLUSTERS_FULFILLED",
  },
]
`;

exports[`mappingWizard actions should fetch target clusters and return PENDING and REJECTED action 1`] = `
Array [
  Object {
    "type": "FETCH_V2V_TARGET_CLUSTERS_PENDING",
  },
  Object {
    "error": true,
    "payload": [Error: Request failed with status code 404],
    "type": "FETCH_V2V_TARGET_CLUSTERS_REJECTED",
  },
]
`;
